First tests

Liljenmaa opened the pull request.

Liljenmaa committed:
Add Coveralls badge to README

Liljenmaa committed:
Add tests for half of the testable files

II-KA commented on dummy.test.js:
```diff
@@ -1,3 +1,11 @@
-test("tests are run", () => {
-  expect(true).toBe(true);
+import countBy from "./src/countBy.js";
+
+test("countBy works", () => {
+  expect(countBy([
+     { 'user': 'barney', 'active': true },
+     { 'user': 'betty', 'active': true },
+     { 'user': 'fred', 'active': false }
+   ], value => value.active)).toBe(
+    { 'true': 2, 'false': 1 }
+  );
 });
```
These tests are duplicated in countBy.test.js. This dummy file can probably be deleted since the test setup is now working.

II-KA commented on test/countBy.test.js:
```diff
@@ -0,0 +1,11 @@
+import countBy from "../src/countBy.js";
+
+test("countBy works", () => {
+    expect(countBy([
+        { 'user': 'barney', 'active': true },
+        { 'user': 'betty', 'active': true },
+        { 'user': 'fred', 'active': false }
+    ], value => value.active)).toBe(
+        { 'true': 2, 'false': 1 }
+    );
+});
```
I'd recommend using describe-it "syntax" to group the tests (in all test files), since we are going to have multiple tests for each utility function. `It()` can be used to create new tests and `describe()` can be used to further group the tests if needed. The top level describe should contain the function name 👍 

Additionally, separating the users array to a variable makes the test easier to read in my opinion 

```suggestion
describe("countBy", () => {
  it("correctly counts active users", () => {
    const users = [
      { user: "barney", active: true },
      { user: "betty", active: true },
      { user: "fred", active: false },
    ];
    expect(countBy(users, (value) => value.active)).toBe({ true: 2, false: 1 });
  });
});
```

II-KA commented on test/map.test.js:
```diff
@@ -0,0 +1,11 @@
+}
+
+describe("map", () => {
+  it("maps numbers to squared", () => {
+    expect(map([4, 8], square)).toIncludeAllMembers([16, 64]);
+  });
```
Looks like `toBe` does not compare arrays correctly. Changed to https://jest-extended.jestcommunity.dev/docs/matchers/Array#toincludeallmembersmembers and the test passes

Liljenmaa replied:
Or should it be "toStrictEqual"?

II-KA replied:
Could also be ok, but since we are using jest-extended lets prefer that. It also has a lot of matchers for object types

II-KA commented:
Btw the command `npx jest <filename>` can be used to run individual test files. Running all of them in development might get annoying since multiple fail

II-KA commented on test/words.test.js:
```diff
@@ -0,0 +1,13 @@
+import words from "../src/words.js";
+
+test("words without regex", () => {
+    expect(words('fred, barney, & pebbles')).toBe(
+        ['fred', 'barney', 'pebbles']
+    );
+});
+
+test("words with regex", () => {
+    expect(words('fred, barney, & pebbles', /[^, ]+/g)).toBe(
+        ['fred', 'barney', '&', 'pebbles']
+    );
+});
```
Using toIncludeAllMembers will make the tests pass

```suggestion
test("words without regex", () => {
    expect(words('fred, barney, & pebbles')).toIncludeAllMembers(
        ['fred', 'barney', 'pebbles']
    );
});

test("words with regex", () => {
    expect(words('fred, barney, & pebbles', /[^, ]+/g)).toIncludeAllMembers(
        ['fred', 'barney', '&', 'pebbles']
    );
});
```

II-KA commented on test/map.test.js:
```diff
@@ -0,0 +1,19 @@
+    it("regards null as empty array", () => {
+        const square = (n) => n * n;
+
+        expect(map(null, square)).toIncludeAllMembers(
+            []
+        );
```
```suggestion
        expect(map(null, square)).toBeArrayOfSize(0)
```

II-KA commented on test/countBy.test.js:
```diff
@@ -0,0 +1,12 @@
+import countBy from "../src/countBy.js";
+
+describe("countBy", () => {
+    it("correctly counts active users", () => {
+        const users = [
+            { 'user': 'barney', 'active': true },
+            { 'user': 'betty', 'active': true },
+            { 'user': 'fred', 'active': false },
+        ];
+        expect(countBy(users, (value) => value.active)).toBe({ true: 2, false: 1 });
+    });
+});
```
I'd recommend adding tests for some corner cases e.g. no user is active

II-KA commented on test/isObjectLike.test.js:
```diff
@@ -0,0 +1,27 @@
+import isObjectLike from "../src/isObjectLike.js";
+
+describe("isObjectLike", () => {
```
How about testing a date?

II-KA commented on test/toNumber.test.js:
```diff
@@ -0,0 +1,146 @@
+import toNumber from "../src/toNumber.js";
+
+describe("toNumber", () => {
```
I'd add more describes inside this to group the `it()` tests

```
describe('transforms to itself', () => {});
describe('transforms to number', () => {});
describe('transforms to not a number', () => {});
...
```

II-KA commented on test/countBy.test.js:
```diff
@@ -0,0 +1,57 @@
+import countBy from "../src/countBy.js";
+
+describe("countBy", () => {
+    it("can correctly count attributes from an array of objects", () => {
+        const users = [
+            { user: 'barney', active: true },
+            { user: 'betty', active: true },
+            { user: 'fred', active: false },
+        ];
```
Users array is used in multiple places, so it could be defined once in the beginning
```suggestion
const users = [
  { user: 'barney', active: true },
  { user: 'betty', active: true },
  { user: 'fred', active: false },
];

describe("countBy", () => {
    it("can correctly count attributes from an array of objects", () => {
```

II-KA committed:
Refactor tests

Group it()-statements with describe() and rename test descriptions
Define commonly used variables in the beginning of the file.

II-KA commented:
Rebased on main
Refactored tests:
- Group it()-statements with describe() and rename test descriptions
- Define commonly used variables in the beginning of the file
